From 65894588b0355350f69774df8cd9873153c8bf27 Mon Sep 17 00:00:00 2001
From: Michael Simacek <michael.simacek@oracle.com>
Date: Thu, 9 Feb 2023 16:33:15 +0100
Subject: [PATCH] Use prebuilt cmake by default

---
 CMakeLists.txt | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7a3a134..89bf75e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -25,7 +25,8 @@ if(CMakePythonDistributions_SUPERBUILD)
 
   #-----------------------------------------------------------------------------
   # Options
-  set(default ON)
+  # GraalPy change
+  set(default OFF)
   if(WIN32 OR APPLE)
     set(default OFF)
   endif()
-- 
2.39.1

