From b1c948a07cfe2ffbb5e0954f0ee87dddd78e63ab Mon Sep 17 00:00:00 2001
From: Michael Simacek <michael.simacek@oracle.com>
Date: Tue, 18 Jan 2022 15:54:16 +0100
Subject: [PATCH] Adapt for GraalPython

---
 setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index c6ee5bf..b398be9 100644
--- a/setup.py
+++ b/setup.py
@@ -52,7 +52,7 @@ def show_message(*lines):
     print("=" * 74)
 
 
-supports_speedups = platform.python_implementation() not in {"PyPy", "Jython"}
+supports_speedups = platform.python_implementation() not in {"PyPy", "Jython", "GraalVM"}
 
 if os.environ.get("CIBUILDWHEEL", "0") == "1" and supports_speedups:
     run_setup(True)
-- 
2.31.1

