diff --git a/pandas/io/common.py b/pandas/io/common.py
index f02c43d..a6df5bc 100644
--- a/pandas/io/common.py
+++ b/pandas/io/common.py
@@ -660,8 +660,8 @@ def get_handle(
 
     # validate encoding and errors
     codecs.lookup(encoding)
-    if isinstance(errors, str):
-        codecs.lookup_error(errors)
+    # if isinstance(errors, str):
+    #     codecs.lookup_error(errors)
 
     # open URLs
     ioargs = _get_filepath_or_buffer(
