diff --git a/pandas/io/common.py b/pandas/io/common.py
index f31de63..1858912 100644
--- a/pandas/io/common.py
+++ b/pandas/io/common.py
@@ -706,8 +706,8 @@ def get_handle(
 
     # validate encoding and errors
     codecs.lookup(encoding)
-    if isinstance(errors, str):
-        codecs.lookup_error(errors)
+    # if isinstance(errors, str):
+    #     codecs.lookup_error(errors)
 
     # open URLs
     ioargs = _get_filepath_or_buffer(
diff --git a/setup.py b/setup.py
index 12e8aa3..0d07514 100755
--- a/setup.py
+++ b/setup.py
@@ -419,7 +419,7 @@ def maybe_cythonize(extensions, *args, **kwargs):
     parser.add_argument("--parallel", "-j", type=int, default=1)
     parsed, _ = parser.parse_known_args()
 
-    kwargs["nthreads"] = parsed.parallel
+    kwargs["nthreads"] = 1 # parsed.parallel
     build_ext.render_templates(_pxifiles)
     return cythonize(extensions, *args, **kwargs)
 
