diff --git a/threadpoolctl.py b/threadpoolctl.py
index 9386882..b380083 100644
--- a/threadpoolctl.py
+++ b/threadpoolctl.py
@@ -403,10 +403,11 @@ class _ThreadpoolInfo():
         c_func_signature = ctypes.CFUNCTYPE(
             ctypes.c_int,  # Return type
             ctypes.POINTER(_dl_phdr_info), ctypes.c_size_t, ctypes.c_char_p)
-        c_match_module_callback = c_func_signature(match_module_callback)
+        # callbacks not yet supported
+        # c_match_module_callback = c_func_signature(match_module_callback)
 
         data = ctypes.c_char_p(b"")
-        libc.dl_iterate_phdr(c_match_module_callback, data)
+        # libc.dl_iterate_phdr(c_match_module_callback, data)
 
     def _find_modules_with_dyld(self):
         """Loop through loaded libraries and return binders on supported ones
