diff -r -u a/_distutils_hack/__init__.py b/_distutils_hack/__init__.py
--- a/_distutils_hack/__init__.py	2022-06-17 21:46:23.000000000 +0200
+++ b/_distutils_hack/__init__.py	2022-07-29 10:37:40.000000000 +0200
@@ -44,7 +44,9 @@
     """
     Allow selection of distutils by environment variable.
     """
-    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'local')
+    # which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'local')
+    # GraalVM change: default to stdlib
+    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'stdlib')
     return which == 'local'


@@ -200,7 +202,8 @@


 def add_shim():
-    DISTUTILS_FINDER in sys.meta_path or insert_shim()
+    if enabled():
+        DISTUTILS_FINDER in sys.meta_path or insert_shim()


 class shim:
@@ -212,7 +215,8 @@


 def insert_shim():
-    sys.meta_path.insert(0, DISTUTILS_FINDER)
+    if enabled():
+        sys.meta_path.insert(0, DISTUTILS_FINDER)


 def remove_shim():
diff -r -u a/setuptools/_vendor/importlib_metadata/_compat.py b/setuptools/_vendor/importlib_metadata/_compat.py
--- a/setuptools/_vendor/importlib_metadata/_compat.py	2022-06-17 21:46:23.000000000 +0200
+++ b/setuptools/_vendor/importlib_metadata/_compat.py	2022-07-29 10:56:09.000000000 +0200
@@ -36,7 +36,7 @@
     def matches(finder):
         return getattr(
             finder, '__module__', None
-        ) == '_frozen_importlib_external' and hasattr(finder, 'find_distributions')
+        ) in ('_frozen_importlib_external', 'importlib._bootstrap_external') and hasattr(finder, 'find_distributions')

     for finder in filter(matches, sys.meta_path):  # pragma: nocover
         del finder.find_distributions
