diff --git a/mesonpy/__init__.py b/mesonpy/__init__.py
index 3ea1325..47ba9ce 100644
--- a/mesonpy/__init__.py
+++ b/mesonpy/__init__.py
@@ -109,7 +109,8 @@ _STYLES = _init_colors()  # holds the color values, should be _COLORS or _NO_COL
 
 _EXTENSION_SUFFIXES = frozenset(importlib.machinery.EXTENSION_SUFFIXES)
 _EXTENSION_SUFFIX_REGEX = re.compile(r'^\.(?:(?P<abi>[^.]+)\.)?(?:so|pyd|dll)$')
-assert all(re.match(_EXTENSION_SUFFIX_REGEX, x) for x in _EXTENSION_SUFFIXES)
+if sys.implementation.name != 'graalpy':
+    assert all(re.match(_EXTENSION_SUFFIX_REGEX, x) for x in _EXTENSION_SUFFIXES)
 
 
 def _showwarning(
