diff --git a/keras/src/datasets/cifar.py b/keras/src/datasets/cifar.py
index 40040f8..65ca7e4 100644
--- a/keras/src/datasets/cifar.py
+++ b/keras/src/datasets/cifar.py
@@ -14,7 +14,7 @@
 # ==============================================================================
 """Utilities common to CIFAR10 and CIFAR100 datasets."""
 
-import _pickle as cPickle
+import pickle as cPickle
 
 
 def load_batch(fpath, label_key="labels"):
diff --git a/keras/src/engine/training.py b/keras/src/engine/training.py
index bcd02a8..d2db9af 100644
--- a/keras/src/engine/training.py
+++ b/keras/src/engine/training.py
@@ -62,7 +62,7 @@ from keras.src.utils import version_utils
 from keras.src.utils.mode_keys import ModeKeys
 
 try:
-    import h5py
+    raise ImportError; import h5py
 except ImportError:
     h5py = None
 
diff --git a/keras/src/saving/legacy/hdf5_format.py b/keras/src/saving/legacy/hdf5_format.py
index 2777763..af903d9 100644
--- a/keras/src/saving/legacy/hdf5_format.py
+++ b/keras/src/saving/legacy/hdf5_format.py
@@ -35,7 +35,7 @@ from keras.src.utils.io_utils import ask_to_proceed_with_overwrite
 from tensorflow.python.platform import tf_logging as logging
 
 try:
-    import h5py
+    raise ImportError; import h5py
 
     HDF5_OBJECT_HEADER_LIMIT = 64512
 except ImportError:
diff --git a/keras/src/saving/legacy/save.py b/keras/src/saving/legacy/save.py
index 98ada0e..a8c6482 100644
--- a/keras/src/saving/legacy/save.py
+++ b/keras/src/saving/legacy/save.py
@@ -31,7 +31,7 @@ from keras.src.utils import io_utils
 from keras.src.utils import traceback_utils
 
 try:
-    import h5py
+    raise ImportError; import h5py
 except ImportError:
     h5py = None
 
diff --git a/keras/src/saving/saving_api.py b/keras/src/saving/saving_api.py
index 15d0eb8..d02579f 100644
--- a/keras/src/saving/saving_api.py
+++ b/keras/src/saving/saving_api.py
@@ -26,7 +26,7 @@ from keras.src.saving.legacy import save as legacy_sm_saving_lib
 from keras.src.utils import io_utils
 
 try:
-    import h5py
+    raise ImportError; import h5py
 except ImportError:
     h5py = None
 
diff --git a/keras/src/saving/saving_lib.py b/keras/src/saving/saving_lib.py
index 8e2bdba..3008068 100644
--- a/keras/src/saving/saving_lib.py
+++ b/keras/src/saving/saving_lib.py
@@ -38,7 +38,7 @@ from keras.src.utils import generic_utils
 from keras.src.utils import io_utils
 
 try:
-    import h5py
+    raise ImportError; import h5py
 except ImportError:
     h5py = None
 
diff --git a/keras/src/testing_infra/test_combinations.py b/keras/src/testing_infra/test_combinations.py
index 7e17589..df39c17 100644
--- a/keras/src/testing_infra/test_combinations.py
+++ b/keras/src/testing_infra/test_combinations.py
@@ -27,7 +27,7 @@ import keras.src as keras
 from keras.src.testing_infra import test_utils
 
 try:
-    import h5py
+    raise ImportError; import h5py
 except ImportError:
     h5py = None
 
