diff --git a/keras/src/datasets/cifar.py b/keras/src/datasets/cifar.py
index 40040f8..65ca7e4 100644
--- a/keras/src/datasets/cifar.py
+++ b/keras/src/datasets/cifar.py
@@ -14,7 +14,7 @@
 # ==============================================================================
 """Utilities common to CIFAR10 and CIFAR100 datasets."""
 
-import _pickle as cPickle
+import pickle as cPickle
 
 
 def load_batch(fpath, label_key="labels"):
