diff --git a/setup.py b/setup.py
index 3a55a00..0c6f89a 100644
--- a/setup.py
+++ b/setup.py
@@ -11,7 +11,8 @@ from skbuild import setup
 # Blosc version
 VERSION = open("VERSION").read().strip()
 # Create the version.py file
-open("blosc2/version.py", "w").write(f'__version__ = "{VERSION}"\n')
+with open("blosc2/version.py", "w") as f:
+    f.write(f'__version__ = "{VERSION}"\n')
 
 
 def exclude_pkgconfig(cmake_manifest):
