From 6c8536c3f09ec5ee435620b51cea2b2072b94ad5 Mon Sep 17 00:00:00 2001
From: Mohaned Qunaibit <mohaned.qunaibit@oracle.com>
Date: Wed, 25 Oct 2023 16:13:02 +0300
Subject: [PATCH] patch h5py 3.10.0

---
 pyproject.toml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pyproject.toml b/pyproject.toml
index ee573d2f..50179ba7 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,7 +1,7 @@
 [build-system]
 requires = [
     "Cython >=0.29.31,<1",
-    "oldest-supported-numpy",
+    "numpy==1.23.5",
     "pkgconfig",
     "setuptools >=61",
 ]
-- 
2.25.1

