diff --git a/joblib/_parallel_backends.py b/joblib/_parallel_backends.py
index 1852b0d..7b99112 100644
--- a/joblib/_parallel_backends.py
+++ b/joblib/_parallel_backends.py
@@ -511,6 +511,10 @@ class MultiprocessingBackend(PoolManagerMixin, AutoBatchingMixin,
                     stacklevel=3)
             return 1
 
+        # GraalPy change:
+        if n_jobs < 0:
+            return max(__graalpython__.get_max_process_count() + 1 + n_jobs, 1)
+
         return super(MultiprocessingBackend, self).effective_n_jobs(n_jobs)
 
     def configure(self, n_jobs=1, parallel=None, prefer=None, require=None,
@@ -592,7 +596,9 @@ class LokyBackend(AutoBatchingMixin, ParallelBackendBase):
                     stacklevel=3)
             return 1
         elif n_jobs < 0:
-            n_jobs = max(cpu_count() + 1 + n_jobs, 1)
+            # GraalPy change:
+            # n_jobs = max(cpu_count() + 1 + n_jobs, 1)
+            n_jobs = max(__graalpython__.get_max_process_count() + 1 + n_jobs, 1)
         return n_jobs
 
     def apply_async(self, func, callback=None):
diff --git a/joblib/externals/cloudpickle/cloudpickle.py b/joblib/externals/cloudpickle/cloudpickle.py
index 317be69..0c23e88 100644
--- a/joblib/externals/cloudpickle/cloudpickle.py
+++ b/joblib/externals/cloudpickle/cloudpickle.py
@@ -512,10 +512,12 @@ def _walk_global_ops(code):
     """
     Yield referenced name for all global-referencing instructions in *code*.
     """
-    for instr in dis.get_instructions(code):
-        op = instr.opcode
-        if op in GLOBAL_OPS:
-            yield instr.argval
+    # GraalPy change: we don't support dis
+    yield from code.co_names
+    # for instr in dis.get_instructions(code):
+    #     op = instr.opcode
+    #     if op in GLOBAL_OPS:
+    #         yield instr.argval
 
 
 def _extract_class_dict(cls):
diff --git a/joblib/externals/loky/backend/fork_exec.py b/joblib/externals/loky/backend/fork_exec.py
index 2353c42..f083a63 100644
--- a/joblib/externals/loky/backend/fork_exec.py
+++ b/joblib/externals/loky/backend/fork_exec.py
@@ -35,9 +35,6 @@ def fork_exec(cmd, keep_fds, env=None):
     env = env or {}
     child_env = {**os.environ, **env}
 
-    pid = os.fork()
-    if pid == 0:  # pragma: no cover
-        close_fds(keep_fds)
-        os.execve(sys.executable, cmd, child_env)
-    else:
-        return pid
+    import subprocess
+    p = subprocess.Popen(cmd, executable=sys.executable, pass_fds=keep_fds, env=child_env)
+    return p.pid
