diff --git a/mypyc_util.h b/mypyc_util.h
index 115c309..d0c825f 100644
--- a/mypyc_util.h
+++ b/mypyc_util.h
@@ -82,7 +82,7 @@
 
 static inline void CPy_INCREF_NO_IMM(PyObject *op)
 {
-    Py_REFCNT(op)++;
+    Py_INCREF(op);
 }
 
 static inline void CPy_DECREF_NO_IMM(PyObject *op)
diff --git a/pythonsupport.h b/pythonsupport.h
index 60454b6..1c38757 100644
--- a/pythonsupport.h
+++ b/pythonsupport.h
@@ -257,8 +257,14 @@ list_resize(PyListObject *self, Py_ssize_t newsize)
         PyErr_NoMemory();
         return -1;
     }
-    PySequence_Fast_ITEMS((PyObject*)self) = items;
+    
+    Py_ssize_t oldsize = Py_SIZE(self);
     Py_SET_SIZE(self, newsize);
+
+    for (Py_ssize_t i = oldsize; i < newsize; i++) {
+        PyList_SetItem(self, i, items[i]);
+    }
+
     self->allocated = new_allocated;
     return 0;
 }
