diff --git a/setupext.py b/setupext.py
index a898d64..1f0b0b4 100644
--- a/setupext.py
+++ b/setupext.py
@@ -631,7 +631,7 @@ class FreeType(SetupPackage):
                     var: value
                     for var, value in sysconfig.get_config_vars().items()
                     if var in {"CC", "CFLAGS", "CXX", "CXXFLAGS", "LD",
-                               "LDFLAGS"}
+                               "LDFLAGS", "AR", "RANLIB"}
                 },
                 **env,
             }
