diff --git a/opencv/cmake/OpenCVFindLibsPerf.cmake b/opencv/cmake/OpenCVFindLibsPerf.cmake
index dfc9459..861a39c 100644
--- a/opencv/cmake/OpenCVFindLibsPerf.cmake
+++ b/opencv/cmake/OpenCVFindLibsPerf.cmake
@@ -84,7 +84,13 @@ if(WITH_EIGEN AND NOT HAVE_EIGEN)
         set(EIGEN_WORLD_VERSION ${EIGEN3_WORLD_VERSION})
         set(EIGEN_MAJOR_VERSION ${EIGEN3_MAJOR_VERSION})
         set(EIGEN_MINOR_VERSION ${EIGEN3_MINOR_VERSION})
-      else()  # Eigen config file
+      elseif(DEFINED Eigen3_VERSION_MAJOR) # Recommended package config variables
+        # see https://github.com/opencv/opencv/issues/27530
+        set(EIGEN_WORLD_VERSION ${Eigen3_VERSION_MAJOR})
+        set(EIGEN_MAJOR_VERSION ${Eigen3_VERSION_MINOR})
+        set(EIGEN_MINOR_VERSION ${Eigen3_VERSION_PATCH})
+      else()  # Deprecated package config variables
+        # Removed on master at https://gitlab.com/libeigen/eigen/-/commit/f2984cd0778dd0a1d7e74216d826eaff2bc6bfab
         set(EIGEN_WORLD_VERSION ${EIGEN3_VERSION_MAJOR})
         set(EIGEN_MAJOR_VERSION ${EIGEN3_VERSION_MINOR})
         set(EIGEN_MINOR_VERSION ${EIGEN3_VERSION_PATCH})
diff --git a/setup.py b/setup.py
index 8941218..be79b06 100755
--- a/setup.py
+++ b/setup.py
@@ -187,7 +187,7 @@ def main():
             "-DBUILD_TESTS=OFF",
             "-DBUILD_PERF_TESTS=OFF",
             "-DBUILD_DOCS=OFF",
-            "-DPYTHON3_LIMITED_API=ON",
+            "-DPYTHON3_LIMITED_API=OFF",
             "-DBUILD_OPENEXR=ON",
         ]
         + (
