diff --git a/pandas/_libs/include/pandas/vendored/klib/khash_python.h b/pandas/_libs/include/pandas/vendored/klib/khash_python.h
index 05cdeba..00b6230 100644
--- a/pandas/_libs/include/pandas/vendored/klib/khash_python.h
+++ b/pandas/_libs/include/pandas/vendored/klib/khash_python.h
@@ -163,13 +163,15 @@ static inline int floatobject_cmp(PyFloatObject *a, PyFloatObject *b) {
 // PyObject_RichCompareBool for complexobjects has a different behavior
 // needs to be replaced
 static inline int complexobject_cmp(PyComplexObject *a, PyComplexObject *b) {
-  return (isnan(a->cval.real) && isnan(b->cval.real) && isnan(a->cval.imag) &&
-          isnan(b->cval.imag)) ||
-         (isnan(a->cval.real) && isnan(b->cval.real) &&
-          a->cval.imag == b->cval.imag) ||
-         (a->cval.real == b->cval.real && isnan(a->cval.imag) &&
-          isnan(b->cval.imag)) ||
-         (a->cval.real == b->cval.real && a->cval.imag == b->cval.imag);
+  Py_complex a_cval = PyComplex_AsCComplex((PyObject*)a);
+  Py_complex b_cval = PyComplex_AsCComplex((PyObject*)b);
+  return (isnan(a_cval.real) && isnan(b_cval.real) && isnan(a_cval.imag) &&
+          isnan(b_cval.imag)) ||
+         (isnan(a_cval.real) && isnan(b_cval.real) &&
+          a_cval.imag == b_cval.imag) ||
+         (a_cval.real == b_cval.real && isnan(a_cval.imag) &&
+          isnan(b_cval.imag)) ||
+         (a_cval.real == b_cval.real && a_cval.imag == b_cval.imag);
 }
 
 static inline int pyobject_cmp(PyObject *a, PyObject *b);
@@ -237,8 +239,9 @@ static inline Py_hash_t floatobject_hash(PyFloatObject *key) {
 
 // replaces _Py_HashDouble with _Pandas_HashDouble
 static inline Py_hash_t complexobject_hash(PyComplexObject *key) {
-  Py_uhash_t realhash = (Py_uhash_t)_Pandas_HashDouble(key->cval.real);
-  Py_uhash_t imaghash = (Py_uhash_t)_Pandas_HashDouble(key->cval.imag);
+  Py_complex cval = PyComplex_AsCComplex((PyObject*)key);
+  Py_uhash_t realhash = (Py_uhash_t)_Pandas_HashDouble(cval.real);
+  Py_uhash_t imaghash = (Py_uhash_t)_Pandas_HashDouble(cval.imag);
   if (realhash == (Py_uhash_t)-1 || imaghash == (Py_uhash_t)-1) {
     return -1;
   }
