diff --git a/pip/__init__.py b/pip/__init__.py
index efefccf..5f5c952 100644
--- a/pip/__init__.py
+++ b/pip/__init__.py
@@ -11,3 +11,6 @@ def main(args: Optional[List[str]] = None) -> int:
     from pip._internal.utils.entrypoints import _wrapper
 
     return _wrapper(args)
+
+
+__GRAALPY_PATCHED = True
diff --git a/pip/_internal/cli/cmdoptions.py b/pip/_internal/cli/cmdoptions.py
index 0b7cff7..07c86bf 100644
--- a/pip/_internal/cli/cmdoptions.py
+++ b/pip/_internal/cli/cmdoptions.py
@@ -357,7 +357,9 @@ def extra_index_url() -> Option:
         dest="extra_index_urls",
         metavar="URL",
         action="append",
-        default=[],
+        # GraalPy change: add default extra index with our prebuilt binaries. If PIP_INDEX_URL is set,
+        # assume the user configured a custom index and does not want to use any default index.
+        default=[] if "PIP_INDEX_URL" in os.environ else ["https://www.graalvm.org/python/wheels/"],
         help="Extra URLs of package indexes to use in addition to "
         "--index-url. Should follow the same rules as "
         "--index-url.",
@@ -892,7 +894,7 @@ disable_pip_version_check: Callable[..., Option] = partial(
     "--disable-pip-version-check",
     dest="disable_pip_version_check",
     action="store_true",
-    default=False,
+    default=True, # GraalPy: we do not want to incentivize the upgrade
     help="Don't periodically check PyPI to determine whether a new version "
     "of pip is available for download. Implied with --no-index.",
 )
diff --git a/pip/_internal/cli/index_command.py b/pip/_internal/cli/index_command.py
index db105d0..2514133 100644
--- a/pip/_internal/cli/index_command.py
+++ b/pip/_internal/cli/index_command.py
@@ -24,6 +24,9 @@ if TYPE_CHECKING:
 
 logger = logging.getLogger(__name__)
 
+# GraalPy change: we need the session to fetch remote metadata, but it would be too difficult to pass it down all
+# the possible code paths
+_GRAALPY_SESSION = None
 
 def _create_truststore_ssl_context() -> Optional["SSLContext"]:
     if sys.version_info < (3, 10):
@@ -78,6 +81,10 @@ class SessionCommandMixin(CommandContextMixIn):
             # automatically ContextManager[Any] and self._session becomes Any,
             # then https://github.com/python/mypy/issues/7696 kicks in
             assert self._session is not None
+            # GraalPy change
+            global _GRAALPY_SESSION
+            _GRAALPY_SESSION = self._session
+
         return self._session
 
     def _build_session(
diff --git a/pip/_internal/index/package_finder.py b/pip/_internal/index/package_finder.py
index 0d65ce3..63aa513 100644
--- a/pip/_internal/index/package_finder.py
+++ b/pip/_internal/index/package_finder.py
@@ -36,6 +36,7 @@ from pip._internal.utils.logging import indent_log
 from pip._internal.utils.misc import build_netloc
 from pip._internal.utils.packaging import check_requires_python
 from pip._internal.utils.unpacking import SUPPORTED_EXTENSIONS
+from graalpy_pip_extensions import apply_graalpy_sort_order, get_graalpy_candidates
 
 if TYPE_CHECKING:
     from pip._vendor.typing_extensions import TypeGuard
@@ -477,6 +478,7 @@ class CandidateEvaluator:
 
         return sorted(filtered_applicable_candidates, key=self._sort_key)
 
+    @apply_graalpy_sort_order
     def _sort_key(self, candidate: InstallationCandidate) -> CandidateSortingKey:
         """
         Function to pass as the `key` argument to a call to sorted() to sort
@@ -845,8 +847,11 @@ class PackageFinder:
 
             logger.debug("Local files found: %s", ", ".join(paths))
 
+        # We add links for few packages that don't publish sdists
+        graalpy_candidates = get_graalpy_candidates(project_name)
+
         # This is an intentional priority ordering
-        return file_candidates + page_candidates
+        return file_candidates + page_candidates + graalpy_candidates
 
     def make_candidate_evaluator(
         self,
diff --git a/pip/_internal/network/download.py b/pip/_internal/network/download.py
index 5c3bce3..f3057d0 100644
--- a/pip/_internal/network/download.py
+++ b/pip/_internal/network/download.py
@@ -7,6 +7,7 @@ import mimetypes
 import os
 from typing import Iterable, Optional, Tuple
 
+from graalpy_pip_extensions import AddedSourceLink
 from pip._vendor.requests.models import Response
 
 from pip._internal.cli.progress_bars import get_download_progress_renderer
@@ -99,7 +100,7 @@ def _get_http_response_filename(resp: Response, link: Link) -> str:
     filename = link.filename  # fallback
     # Have a look at the Content-Disposition header for a better guess
     content_disposition = resp.headers.get("content-disposition")
-    if content_disposition:
+    if content_disposition and not isinstance(link, AddedSourceLink):
         filename = parse_content_disposition(content_disposition, filename)
     ext: Optional[str] = splitext(filename)[1]
     if not ext:
diff --git a/pip/_internal/operations/install/wheel.py b/pip/_internal/operations/install/wheel.py
index aef42aa..0dcc357 100644
--- a/pip/_internal/operations/install/wheel.py
+++ b/pip/_internal/operations/install/wheel.py
@@ -590,6 +590,9 @@ def _install_wheel(  # noqa: C901, PLR0915 function is too long
         file.save()
         record_installed(file.src_record_path, file.dest_path, file.changed)
 
+    from graalpy_pip_extensions import apply_graalpy_patches
+    apply_graalpy_patches(wheel_path, lib_dir)
+
     def pyc_source_file_paths() -> Generator[str, None, None]:
         # We de-duplicate installation paths, since there can be overlap (e.g.
         # file in .data maps to same location as file in wheel root).
diff --git a/pip/_internal/resolution/resolvelib/candidates.py b/pip/_internal/resolution/resolvelib/candidates.py
index 6617644..ad52082 100644
--- a/pip/_internal/resolution/resolvelib/candidates.py
+++ b/pip/_internal/resolution/resolvelib/candidates.py
@@ -23,6 +23,7 @@ from pip._internal.req.constructors import (
 from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.direct_url_helpers import direct_url_from_link
 from pip._internal.utils.misc import normalize_version_info
+import graalpy_pip_extensions as graalpy
 
 from .base import Candidate, Requirement, format_name
 
@@ -253,6 +254,8 @@ class _InstallRequirementBackedCandidate(Candidate):
         for r in requires:
             yield from self._factory.make_requirements_from_spec(str(r), self._ireq)
         yield self._factory.make_requires_python_requirement(self.dist.requires_python)
+        if self.name == 'virtualenv' and not graalpy.DISABLE_PATCHING:
+            yield from self._factory.make_requirements_from_spec('graalpy-virtualenv-seeder', self._ireq)
 
     def get_install_requirement(self) -> Optional[InstallRequirement]:
         return self._ireq
diff --git a/pip/_internal/utils/unpacking.py b/pip/_internal/utils/unpacking.py
index 875e30e..f6562cf 100644
--- a/pip/_internal/utils/unpacking.py
+++ b/pip/_internal/utils/unpacking.py
@@ -335,3 +335,5 @@ def unpack_file(
             content_type,
         )
         raise InstallationError(f"Cannot determine archive format of {location}")
+    from graalpy_pip_extensions import apply_graalpy_patches
+    apply_graalpy_patches(filename, location)
diff --git a/pip/_internal/wheel_builder.py b/pip/_internal/wheel_builder.py
index 93f8e1f..32fd4ab 100644
--- a/pip/_internal/wheel_builder.py
+++ b/pip/_internal/wheel_builder.py
@@ -7,6 +7,7 @@ import re
 import shutil
 from typing import Iterable, List, Optional, Tuple
 
+import graalpy_pip_extensions as graalpy
 from pip._vendor.packaging.utils import canonicalize_name, canonicalize_version
 from pip._vendor.packaging.version import InvalidVersion, Version
 
@@ -115,6 +116,10 @@ def _should_cache(
     if _contains_egg_info(base):
         return True
 
+    # GraalPy change
+    if isinstance(req.link, graalpy.AddedSourceLink):
+        return True
+
     # Otherwise, do not cache.
     return False
 
@@ -247,6 +252,11 @@ def _build_one_inside_env(
             )
 
         if wheel_path is not None:
+            # GraalPy change: watermark wheels that we built so that we don't try to patch them when installing
+            # Note: we currently don't patch source builds from SCM links, so we don't watermark them so that they can
+            # get patched with a wheel patch
+            if not req.link.is_vcs:
+                graalpy.mark_wheel(wheel_path)
             wheel_name = os.path.basename(wheel_path)
             dest_path = os.path.join(output_dir, wheel_name)
             try:
diff --git a/pip/_vendor/platformdirs/api.py b/pip/_vendor/platformdirs/api.py
index c50caa6..a46cac7 100644
--- a/pip/_vendor/platformdirs/api.py
+++ b/pip/_vendor/platformdirs/api.py
@@ -80,7 +80,7 @@ class PlatformDirsABC(ABC):  # noqa: PLR0904
     def _append_app_name_and_version(self, *base: str) -> str:
         params = list(base[1:])
         if self.appname:
-            params.append(self.appname)
+            params.append(f'{self.appname}-graalpy')
             if self.version:
                 params.append(self.version)
         path = os.path.join(base[0], *params)  # noqa: PTH118
