diff --git a/setup.py b/setup.py
index a711e24..84d3aea 100644
--- a/setup.py
+++ b/setup.py
@@ -125,7 +125,7 @@ if "--no_ext" in sys.argv or os.environ.get("NO_EXT"):
     except ValueError:
         pass
     ext_modules = []
-elif sys.platform.startswith("java") or sys.platform == "cli" or "PyPy" in sys.version:
+elif sys.platform.startswith("java") or sys.platform == "cli" or "PyPy" in sys.version or sys.implementation.name == 'graalpy':
     sys.stdout.write(
         """
 *****************************************************\n
