diff -ur a/charset_normalizer/utils.py b/charset_normalizer/utils.py
--- a/charset_normalizer/utils.py
+++ b/charset_normalizer/utils.py
@@ -13,7 +13,10 @@
 from re import findall
 from typing import Generator, List, Optional, Set, Tuple, Union

-from _multibytecodec import MultibyteIncrementalDecoder
+try:
+    from _multibytecodec import MultibyteIncrementalDecoder  # type: ignore
+except ImportError:
+    MultibyteIncrementalDecoder = None

 from .constant import (
     ENCODING_MARKS,
@@ -261,7 +264,7 @@
     } or issubclass(
         importlib.import_module("encodings.{}".format(name)).IncrementalDecoder,
         MultibyteIncrementalDecoder,
-    )
+    ) if MultibyteIncrementalDecoder else False


 def identify_sig_or_bom(sequence: bytes) -> Tuple[Optional[str], bytes]:
