diff --git a/bindings/python/Cargo.toml b/bindings/python/Cargo.toml
index 51ac712..a7d1e9b 100644
--- a/bindings/python/Cargo.toml
+++ b/bindings/python/Cargo.toml
@@ -14,8 +14,8 @@ serde = { version = "1.0", features = [ "rc", "derive" ]}
 serde_json = "1.0"
 libc = "0.2"
 env_logger = "0.10.0"
-pyo3 = { version = "0.19" }
-numpy = "0.19.0"
+pyo3 = { git = "https://github.com/timfel/pyo3.git", branch = "v0.19.2.graalpy" }
+numpy = { git = "https://github.com/timfel/rust-numpy.git", branch = "v0.19.graalpy" }
 ndarray = "0.15"
 onig = { version = "6.4", default-features = false }
 itertools = "0.11"
@@ -26,7 +26,7 @@ path = "../../tokenizers"
 
 [dev-dependencies]
 tempfile = "3.8"
-pyo3 = { version = "0.19", features = ["auto-initialize"] }
+pyo3 = { git = "https://github.com/timfel/pyo3.git", branch = "v0.19.2.graalpy", features = ["auto-initialize"] }
 
 [features]
 defaut = ["pyo3/extension-module"]
diff --git a/bindings/python/src/lib.rs b/bindings/python/src/lib.rs
index 8625944..a8ac19d 100644
--- a/bindings/python/src/lib.rs
+++ b/bindings/python/src/lib.rs
@@ -50,14 +50,16 @@ extern "C" fn child_after_fork() {
 pub fn tokenizers(_py: Python, m: &PyModule) -> PyResult<()> {
     let _ = env_logger::try_init_from_env("TOKENIZERS_LOG");
 
+    // GraalPy change: Disable the atfork warning. This triggers a ton of false positives when
+    // jline calls stty and we don't support fork anyway
     // Register the fork callback
-    #[cfg(target_family = "unix")]
-    unsafe {
-        if !REGISTERED_FORK_CALLBACK {
-            libc::pthread_atfork(None, None, Some(child_after_fork));
-            REGISTERED_FORK_CALLBACK = true;
-        }
-    }
+    // #[cfg(target_family = "unix")]
+    // unsafe {
+    //     if !REGISTERED_FORK_CALLBACK {
+    //         libc::pthread_atfork(None, None, Some(child_after_fork));
+    //         REGISTERED_FORK_CALLBACK = true;
+    //     }
+    // }
 
     m.add_class::<tokenizer::PyTokenizer>()?;
     m.add_class::<tokenizer::PyAddedToken>()?;
