diff --git a/charset_normalizer/utils.py b/charset_normalizer/utils.py
index dcb14df..12079e3 100644
--- a/charset_normalizer/utils.py
+++ b/charset_normalizer/utils.py
@@ -11,7 +11,10 @@ from functools import lru_cache
 from re import findall
 from typing import List, Optional, Set, Tuple, Union
 
-from _multibytecodec import MultibyteIncrementalDecoder  # type: ignore
+try:
+    from _multibytecodec import MultibyteIncrementalDecoder  # type: ignore
+except ImportError:
+    MultibyteIncrementalDecoder = None
 
 from .constant import (
     ENCODING_MARKS,
@@ -244,7 +247,7 @@ def is_multi_byte_encoding(name: str) -> bool:
     } or issubclass(
         importlib.import_module("encodings.{}".format(name)).IncrementalDecoder,  # type: ignore
         MultibyteIncrementalDecoder,
-    )
+    ) if MultibyteIncrementalDecoder else False
 
 
 def identify_sig_or_bom(sequence: bytes) -> Tuple[Optional[str], bytes]:
-- 
2.31.1

