From d4baac460f19ec58fc551ed1e7cf1ce0c4723fd9 Mon Sep 17 00:00:00 2001
From: Ondrej Tethal <ondrej.tethal@oracle.com>
Date: Fri, 13 Jan 2023 11:04:00 +0100
Subject: [PATCH] Patch import _datetime

---
 pydriller/domain/commit.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pydriller/domain/commit.py b/pydriller/domain/commit.py
index ba2c6e5..a2188af 100644
--- a/pydriller/domain/commit.py
+++ b/pydriller/domain/commit.py
@@ -18,7 +18,7 @@ Commit, Modification,
 ModificationType and Method.
 """
 import logging
-from _datetime import datetime
+from datetime import datetime
 from enum import Enum
 from pathlib import Path
 from typing import List, Set, Dict, Tuple, Optional
-- 
2.25.1

