diff --git a/python/google/protobuf/internal/api_implementation.py b/python/google/protobuf/internal/api_implementation.py
index 74586487a..10ed520e2 100644
--- a/python/google/protobuf/internal/api_implementation.py
+++ b/python/google/protobuf/internal/api_implementation.py
@@ -96,6 +96,11 @@ if 'PyPy' in sys.version and _implementation_type == 'cpp':
                 'Falling back to the python implementation.')
   _implementation_type = 'python'
 
+if sys.implementation.name == 'graalpy' and _implementation_type == 'cpp':
+  warnings.warn('GraalPy does not work yet with cpp protocol buffers. '
+                'Falling back to the python implementation.')
+  _implementation_type = 'python'
+
 _c_module = None
 
 if _implementation_type == 'cpp':
diff --git a/python/google/protobuf/internal/well_known_types_test.py b/python/google/protobuf/internal/well_known_types_test.py
index a32459a9e..334f0d020 100644
--- a/python/google/protobuf/internal/well_known_types_test.py
+++ b/python/google/protobuf/internal/well_known_types_test.py
@@ -51,6 +51,8 @@ from google.protobuf import text_format
 from google.protobuf.internal import _parameterized
 
 try:
+  # module 'tzdata' not implemented yet in GraalPy [GR-41370]
+  raise ImportError 
   # New module in Python 3.9:
   import zoneinfo  # pylint:disable=g-import-not-at-top
   _TZ_JAPAN = zoneinfo.ZoneInfo('Japan')
