diff --git a/cmake/CMakeLists.txt b/cmake/CMakeLists.txt
index a5d28fb..25b8ae4 100644
--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -33,6 +33,8 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 enable_testing()
 include(Dart)
 
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Wno-error=array-bounds -Wno-error=maybe-uninitialized -Wno-error=uninitialized -Wno-error=restrict -Wno-error=unused-but-set-variable")
+
 if (NOT CMAKE_BUILD_TYPE)
   message(STATUS "Build type not set - using RelWithDebInfo")
   set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Choose build type: Debug Release RelWithDebInfo MinSizeRel." FORCE)
diff --git a/pyproject.toml b/pyproject.toml
index 6bec8bf..51fb108 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,3 +1,7 @@
+[build-system]
+requires = ["setuptools >= 40.6.0", "wheel", "packaging", "numpy"]
+build-backend = "setuptools.build_meta:__legacy__"
+
 [tool.black]
 line-length = 120
 # extend-exclude needs to be a regular expression
diff --git a/requirements.txt b/requirements.txt
new file mode 100644
index 0000000..a9c1327
--- /dev/null
+++ b/requirements.txt
@@ -0,0 +1,6 @@
+coloredlogs
+flatbuffers
+numpy
+packaging
+protobuf
+sympy
diff --git a/setup.py b/setup.py
index 294b975..de3cc48 100644
--- a/setup.py
+++ b/setup.py
@@ -25,6 +25,20 @@ wheel_name_suffix = None
 logger = logging.getLogger()
 
 
+import os
+builddir = os.path.abspath('build/Release')
+if True or not os.path.exists(builddir):
+    build_cmd = [sys.executable, 'tools/ci_build/build.py', '--build_dir', 'build', '--skip_submodule_sync', '--skip_tests', '--config', 'Release', '--enable_pybind']
+    subprocess.check_call(build_cmd)
+    # The cmake build downloads a bunch of sources that need to be patched
+    subprocess.check_call([sys.executable, '-m', 'autopatch_capi', '.'])
+    # Rerun to rebuild the patched sources
+    subprocess.check_call(build_cmd)
+oldcwd = os.path.abspath(os.getcwd())
+os.chdir(builddir)
+
+
+
 def parse_arg_remove_boolean(argv, arg_name):
     arg_value = False
     if arg_name in sys.argv:
diff --git a/tools/ci_build/build.py b/tools/ci_build/build.py
index ed29876..8faf81a 100644
--- a/tools/ci_build/build.py
+++ b/tools/ci_build/build.py
@@ -1129,6 +1129,11 @@ def generate_build_tree(
             "-Ddml_EXTERNAL_PROJECT=ON",
         ]
 
+    if not args.test:
+        cmake_args += [
+            "-Donnxruntime_BUILD_UNIT_TESTS=OFF",
+        ]
+
     if args.use_gdk:
         cmake_args += [
             "-DCMAKE_TOOLCHAIN_FILE=" + os.path.join(source_dir, "cmake", "gdk_toolchain.cmake"),
