diff -u -r a/setupext.py b/setupext.py
--- a/setupext.py
+++ b/setupext.py
@@ -627,7 +627,7 @@
                     var: value
                     for var, value in sysconfig.get_config_vars().items()
                     if var in {"CC", "CFLAGS", "CXX", "CXXFLAGS", "LD",
-                               "LDFLAGS"}
+                               "LDFLAGS", "AR", "RANLIB"}
                 },
                 **env,
             }
