diff --git a/_pytest/_code/code.py b/_pytest/_code/code.py
index 219ebb68f..b0d093f0e 100644
--- a/_pytest/_code/code.py
+++ b/_pytest/_code/code.py
@@ -224,15 +224,17 @@ class TracebackEntry:
             if key is not None:
                 astnode = astcache.get(key, None)
         start = self.getfirstlinesource()
-        try:
-            astnode, _, end = getstatementrange_ast(
-                self.lineno, source, astnode=astnode
-            )
-        except SyntaxError:
-            end = self.lineno + 1
-        else:
-            if key is not None:
-                astcache[key] = astnode
+        end = -1
+        # GraalPython: no support for the ast module so the source cannot be retrieved correctly
+        # try:
+        #     astnode, _, end = getstatementrange_ast(
+        #         self.lineno, source, astnode=astnode
+        #     )
+        # except SyntaxError:
+        #     end = self.lineno + 1
+        # else:
+        #     if key is not None:
+        #         astcache[key] = astnode
         return source[start:end]
 
     source = property(getsource)
diff --git a/_pytest/assertion/__init__.py b/_pytest/assertion/__init__.py
index 64d2267e7..760df1e49 100644
--- a/_pytest/assertion/__init__.py
+++ b/_pytest/assertion/__init__.py
@@ -28,7 +28,7 @@ def pytest_addoption(parser: Parser) -> None:
         action="store",
         dest="assertmode",
         choices=("rewrite", "plain"),
-        default="rewrite",
+        default="plain",
         metavar="MODE",
         help=(
             "Control assertion debugging tools.\n"
 
