diff --git a/_pytest/_code/code.py b/_pytest/_code/code.py
index 7d72234e7..1f0ca7f7e 100644
--- a/_pytest/_code/code.py
+++ b/_pytest/_code/code.py
@@ -222,15 +222,17 @@ class TracebackEntry:
             if key is not None:
                 astnode = astcache.get(key, None)
         start = self.getfirstlinesource()
-        try:
-            astnode, _, end = getstatementrange_ast(
-                self.lineno, source, astnode=astnode
-            )
-        except SyntaxError:
-            end = self.lineno + 1
-        else:
-            if key is not None:
-                astcache[key] = astnode
+        end = -1
+        # GraalPython: no support for the ast module so the source cannot be retrieved correctly
+        # try:
+        #     astnode, _, end = getstatementrange_ast(
+        #         self.lineno, source, astnode=astnode
+        #     )
+        # except SyntaxError:
+        #     end = self.lineno + 1
+        # else:
+        #     if key is not None:
+        #         astcache[key] = astnode
         return source[start:end]

     source = property(getsource)
diff --git a/_pytest/assertion/__init__.py b/_pytest/assertion/__init__.py
index 126929b6a..8ece37d23 100644
--- a/_pytest/assertion/__init__.py
+++ b/_pytest/assertion/__init__.py
@@ -15,7 +15,7 @@ def pytest_addoption(parser):
         action="store",
         dest="assertmode",
         choices=("rewrite", "plain"),
-        default="rewrite",
+        default="plain",
         metavar="MODE",
         help="""Control assertion debugging tools.  'plain'
                             performs no assertion debugging.  'rewrite'

