diff --git a/src/_pytest/_code/code.py b/src/_pytest/_code/code.py
index 4230693..1c9bab8 100644
--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -248,15 +248,17 @@ class TracebackEntry:
             if key is not None:
                 astnode = astcache.get(key, None)
         start = self.getfirstlinesource()
-        try:
-            astnode, _, end = getstatementrange_ast(
-                self.lineno, source, astnode=astnode
-            )
-        except SyntaxError:
-            end = self.lineno + 1
-        else:
-            if key is not None:
-                astcache[key] = astnode
+        end = -1
+        # GraalPython: no support for the ast module so the source cannot be retrieved correctly
+        # try:
+        #     astnode, _, end = getstatementrange_ast(
+        #         self.lineno, source, astnode=astnode
+        #     )
+        # except SyntaxError:
+        #     end = self.lineno + 1
+        # else:
+        #     if key is not None:
+        #         astcache[key] = astnode
         return source[start:end]

     source = property(getsource)
diff --git a/src/_pytest/assertion/__init__.py b/src/_pytest/assertion/__init__.py
index a18cf19..bd7523c 100644
--- a/src/_pytest/assertion/__init__.py
+++ b/src/_pytest/assertion/__init__.py
@@ -26,7 +26,8 @@ def pytest_addoption(parser: Parser) -> None:
         action="store",
         dest="assertmode",
         choices=("rewrite", "plain"),
-        default="rewrite",
+        # default="rewrite",
+        default="plain",
         metavar="MODE",
         help=(
             "Control assertion debugging tools.\n"

