diff --git a/wheel/bdist_wheel.py b/wheel/bdist_wheel.py
index 28a9050..21df4d9 100644
--- a/wheel/bdist_wheel.py
+++ b/wheel/bdist_wheel.py
@@ -13,6 +13,7 @@ import stat
 import sys
 import sysconfig
 import warnings
+import struct
 from collections import OrderedDict
 from email.generator import BytesGenerator, Generator
 from email.policy import EmailPolicy
@@ -66,7 +67,7 @@ def get_platform(archive_root):
     result = sysconfig.get_platform()
     if result.startswith("macosx") and archive_root is not None:
         result = calculate_macosx_platform_tag(archive_root, result)
-    elif result == "linux-x86_64" and sys.maxsize == 2147483647:
+    elif result == "linux-x86_64" and struct.calcsize("P") == 4:
         # pip pull request #3497
         result = "linux-i686"
 
@@ -114,6 +115,9 @@ def get_abi_tag():
         # we want something like pypy36-pp73
         abi = "-".join(soabi.split("-")[:2])
         abi = abi.replace(".", "_").replace("-", "_")
+    elif soabi and impl == "graalpy":
+        abi = "-".join(soabi.split("-")[:3])
+        abi = abi.replace(".", "_").replace("-", "_")
     elif soabi:
         abi = soabi.replace(".", "_").replace("-", "_")
     else:
@@ -297,7 +301,7 @@ class bdist_wheel(Command):
 
             if (
                 plat_name in ("linux-x86_64", "linux_x86_64")
-                and sys.maxsize == 2147483647
+                and struct.calcsize("P") == 4
             ):
                 plat_name = "linux_i686"
 
diff --git a/wheel/vendored/packaging/tags.py b/wheel/vendored/packaging/tags.py
index 19ccbde..dd01a7d 100644
--- a/wheel/vendored/packaging/tags.py
+++ b/wheel/vendored/packaging/tags.py
@@ -4,6 +4,7 @@
 
 import logging
 import platform
+import struct
 import subprocess
 import sys
 import sysconfig
@@ -37,7 +38,7 @@ INTERPRETER_SHORT_NAMES: Dict[str, str] = {
 }
 
 
-_32_BIT_INTERPRETER = sys.maxsize <= 2**32
+_32_BIT_INTERPRETER = struct.calcsize("P") == 4
 
 
 class Tag:
