diff --git a/wheel/pep425tags.py b/wheel/pep425tags.py
index 0c25763..408b407 100644
--- a/wheel/pep425tags.py
+++ b/wheel/pep425tags.py
@@ -38,6 +38,8 @@ def get_abbr_impl():
         return 'ip'
     elif impl == 'CPython':
         return 'cp'
+    elif impl == 'GraalVM':
+        return 'graalpy'
 
     raise LookupError('Unknown Python implementation: ' + impl)
 
@@ -97,6 +99,9 @@ def get_abi_tag():
         abi = '%s%s%s%s%s' % (impl, get_impl_ver(), d, m, u)
     elif soabi and soabi.startswith('cpython-'):
         abi = 'cp' + soabi.split('-')[1]
+    elif soabi and impl == "graalpy":
+        abi = "-".join(soabi.split("-")[:3])
+        abi = abi.replace(".", "_").replace("-", "_")
     elif soabi:
         abi = soabi.replace('.', '_').replace('-', '_')
     else:
