diff --git a/_distutils_hack/__init__.py b/_distutils_hack/__init__.py
index c6f7de6..194293b 100644
--- a/_distutils_hack/__init__.py
+++ b/_distutils_hack/__init__.py
@@ -40,7 +40,9 @@ def enabled():
     """
     Allow selection of distutils by environment variable.
     """
-    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'local')
+    # which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'local')
+    # GraalVM change: default to stdlib
+    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'stdlib')
     return which == 'local'


@@ -163,7 +165,8 @@ DISTUTILS_FINDER = DistutilsMetaFinder()


 def add_shim():
-    DISTUTILS_FINDER in sys.meta_path or insert_shim()
+    if enabled():
+        DISTUTILS_FINDER in sys.meta_path or insert_shim()


 class shim:
@@ -175,7 +178,8 @@ class shim:


 def insert_shim():
-    sys.meta_path.insert(0, DISTUTILS_FINDER)
+     if enabled():
+        sys.meta_path.insert(0, DISTUTILS_FINDER)


 def remove_shim():
diff --git a/setuptools/_vendor/importlib_metadata/_compat.py b/setuptools/_vendor/importlib_metadata/_compat.py
index ef3136f..612377b 100644
--- a/setuptools/_vendor/importlib_metadata/_compat.py
+++ b/setuptools/_vendor/importlib_metadata/_compat.py
@@ -36,7 +36,7 @@ def disable_stdlib_finder():
     def matches(finder):
         return getattr(
             finder, '__module__', None
-        ) == '_frozen_importlib_external' and hasattr(finder, 'find_distributions')
+        ) in ('_frozen_importlib_external', 'importlib._bootstrap_external') and hasattr(finder, 'find_distributions')

     for finder in filter(matches, sys.meta_path):  # pragma: nocover
         del finder.find_distributions
