From a7c84fe6cb196c9b44fd997408b686743fbcc71d Mon Sep 17 00:00:00 2001
From: Michael Simacek <michael.simacek@oracle.com>
Date: Tue, 25 Jan 2022 09:34:51 +0100
Subject: [PATCH] Use stdlib distutils on GraalVM

---
 _distutils_hack/__init__.py | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/_distutils_hack/__init__.py b/_distutils_hack/__init__.py
index 4745f8b9..66d68fa9 100644
--- a/_distutils_hack/__init__.py
+++ b/_distutils_hack/__init__.py
@@ -43,7 +43,8 @@ def enabled():
     """
     Allow selection of distutils by environment variable.
     """
-    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'local')
+    # GraalVM change: default to stdlib
+    which = os.environ.get('SETUPTOOLS_USE_DISTUTILS', 'stdlib')
     return which == 'local'
 
 
@@ -153,7 +154,8 @@ DISTUTILS_FINDER = DistutilsMetaFinder()
 
 
 def add_shim():
-    DISTUTILS_FINDER in sys.meta_path or insert_shim()
+    if enabled():
+        DISTUTILS_FINDER in sys.meta_path or insert_shim()
 
 
 @contextlib.contextmanager
@@ -166,7 +168,8 @@ def shim():
 
 
 def insert_shim():
-    sys.meta_path.insert(0, DISTUTILS_FINDER)
+    if enabled():
+        sys.meta_path.insert(0, DISTUTILS_FINDER)
 
 
 def remove_shim():
-- 
2.31.1

