diff --git a/c/_cffi_backend.c b/c/_cffi_backend.c
index 2cdb76b..ab91032 100644
--- a/c/_cffi_backend.c
+++ b/c/_cffi_backend.c
@@ -5828,7 +5828,7 @@ static CTypeDescrObject *fb_prepare_ctype(struct funcbuilder_s *fb,
     fb->bufferp = NULL;
     fb->fct = NULL;
 
-    pfargs = (CTypeDescrObject **)&PyTuple_GET_ITEM(fargs, 0);
+    pfargs = (CTypeDescrObject **)PySequence_Fast_ITEMS(fargs);
     nargs = PyTuple_GET_SIZE(fargs);
 #if defined(MS_WIN32) && !defined(_WIN64)
     if (fabi == FFI_STDCALL)
